New module: add docker_stack module (cloud/docker/docker_stack)

dariko opened the pull request:
##### SUMMARY
This PR adds a new module, docker_stack, which can be used to manage docker stacks on a docker swarm. 

This module does not use the function/classes defined is [docker_common.py](/ansible/ansible/blob/devel/lib/ansible/module_utils/docker_common.py), as docker stack is not implemented in the docker API but in the docker CLI.

##### ISSUE TYPE
 - New Module Pull Request

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 5cb5228cde) last updated 2017/05/14 08:47:37 (GMT +200)
  config file = 
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/src/ansible/lib/ansible
  executable location = /home/user/src/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATIONS

###### module parameters

| parameter name | required | default | description|
|----------|:-------------:|-------|----|
|name|yes||stack name|
|state|yes||stack status (`present` or `absent`)|
|prune|no||If true will add the `--prune` option to the docker command.|
|compose_yaml|no||string containing the yaml definition of the stack
|compose_file|no||path on the module executing node of a file containing. 


###### return value
The module returns a variable `docker_stack_spec_diff` which contains
the differences between the 'Spec' field of the stack services before
and after applying the new stack definition.


For example, removing the environment variables from a service
named `test_service` in a stack named `test_stack` with this task:

```
-   name: test service
    run_once: yes
    docker_stack:
        state: present
        name: test_stack
        compose_yaml:
            version: '3'
            services:
                test_service:
                    image: alpine:latest
```

will return:
```
changed: [test_host] => {
    "changed": true,
    "docker_stack_specs_diff": "{'test_stack_test_service': {u'TaskTemplate': {u'ContainerSpec': {delete: [u'Env']}}}}",
    "invocation": {
        "module_args": {
            "compose_file": null,
            "compose_yaml": "{'services': {'test_service': {'image': 'alpine:latest'}}, 'version': '3'}",
            "name": "test_stack",
            "state": "present"
        }
    }
}
```

###### requirements
python module jsondiff

dariko committed:
add docker_stack module

dariko committed:
pep8 compliance

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+        default: ""
+        description:
+        -   >
+            Path on the module executing node of a file containing
```
Could you please re-phrase this, IMHO I am not able understand this.

dariko replied:
you're right, now I think now it's more clear

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+            'compose_yaml': dict(),
+            'compose_file': dict(),
+            'prune': dict(default=False, type='bool'),
+            'state': dict(default='present', choices=['present', 'absent'])
```
state in `DOCUMENTATION` say required but I see this is not `required=True` here.

dariko replied:
removed `required: True` from docs

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+    name = module.params['name']
+
+    if state == 'present':
+        if compose_yaml and compose_file:
```
use `required_one_of=[['compose_yaml', 'compose_file']],`

dariko replied:
Using the `required_one_of` would enforce the parameters presence even if `state==absent`, while one of them is required only if `state==present`.

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+                                 "parameters given")
+        elif compose_yaml:
+            fd, compose_file = mkstemp()
+            stack_file = open(compose_file, 'w')
```
use `with` statement

dariko replied:
Ok, I also enclosed the whole part of code which uses the temporary file in a try-finally block, to ensure removal of the temporary file

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+            os.close(fd)
+        elif not compose_file:
+            module.fail_json(msg="compose_yaml or compose_file " +
+                                 "parameters required if state==up")
```
This `state` is not mentioned in `DOCUMENTATION` or `argument_spec`

dariko replied:
It was a leftover from a previous code iteration, corrected to `if state==present`

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+            if isinstance(before_after_differences[k], dict):
+                before_after_differences[k].pop('UpdatedAt', None)
+                before_after_differences[k].pop('Version', None)
+                if len(before_after_differences[k].keys()) == 0:
```
`if not list(before_after_differences[k].keys()):`

dariko replied:
done

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+                if len(before_after_differences[k].keys()) == 0:
+                    before_after_differences.pop(k)
+
+        if before_after_differences == {}:
```
`if not before_after_differences:`

dariko replied:
done

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+                                           "down",
+                                           name])
+
+        if not rc == 0:
```
`if rc != 0:`

dariko replied:
done

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+def main():
+    module = AnsibleModule(
+        argument_spec={
+            'name': dict(required=True),
```
Add type as 'str' for name

dariko replied:
done

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+    docker_stack:
+        state: present
+        name: stack1
+        compose_file: /opt/stack.compose
```
Add blank line after each example

dariko replied:
done

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+        return None
+    else:
+        ret = json.loads(out)[0]['Spec']
+        return(ret)
```
Remove redundant parens

dariko replied:
done

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,226 @@
+                                       "--format",
+                                       "{{.Name}}"])
+    if out != ("Nothing found in stack %s\n" % stack_name):
+        return(out.strip().split('\n'))
```
remove redundant parens

dariko replied:
done

dariko committed:
state is not required

dariko committed:
rephrase compose_file parameter description

dariko committed:
newlines between examples

dariko committed:
remove redundant parens

dariko committed:
specify parameter type for `name`

dariko committed:
style changes

dariko committed:
ensure temporary file removal, wrap temporary file write in with statement

dariko committed:
use mutually_exclusive for compose_file, compose_yaml parameters

dariko commented:
@Akasurde I think i covered all your points.
Instead of `required_one_of` for the `compose_file` and `compose_yaml` parameters I used AnsibleModule's `mutually_exclusive` parameter, which I think it's a better fit it this situation

ulope commented:
In the meantime [docker-py has gained support for swarm mode](https://docker-py.readthedocs.io/en/stable/swarm.html) - so this PR should probably be updated to use that instead of the CLI

dariko commented:
@ulope sorry for the (way) late reply. I don't see anything on the docs you linked about stacks, it looks like code which can be used to join/leave/init swarms.
As far as I know `docker stack` is implemented in the docker client, not in the server..

ulope commented:
@dariko You are correct that the "stack" itself is a function of the client but [services are supported in the API](https://docker-py.readthedocs.io/en/stable/services.html) which basically is all you need to build stack on top of ansible yourself.

dariko commented:
@ulope I submitted #19229 some time ago, which adds a docker_service module: is this what you're looking for? It does not use `docker-py`'s service interface, which was lacking at the time.

Since I submitted this I'm not using that module anymore: this one is concise (*way* less prone to errors) and thanks to the yaml format used by `docker stack` presents a nice and clean interface.
Supporting more docker API versions in particular was growing complexity in the `docker_service` module, while this one simply delegates everything to the docker client (so the `version` attribute in the stack yaml can be used to have consistent results).

All in all I (personally) feel more confident maintaining even a single service with this module (as a one-service stack) than with the `docker_service` one, and I won't suggest using it anymore (if not for some edge cases where the `docker stack` interface might be lacking).

gesellix commented:
What's missing here to get the module added to Ansible? I'm highly interested in using Docker Swarm/Stack via Ansible.

smessmer commented:
+1

ahwebd commented:
Got error:
`ImportError: No module named jsondiff`

dariko commented:
Thanks, the `jsondiff` requirement was already explicit in the module metadata, but I added a note in this PR first message.

chouseknecht commented:
Looks good to me.

+1

jctanner commented:
@Akasurde can you clear your review on this?

bot_status

LuisGerardoAngel commented:
Is this up for next release?

gesellix commented:
@Akasurde can you please check whether your requested changes have been incorporated and clear your review?

dariko commented:
@Akasurde I think I have addressed the problems you mentioned, have i missed something?

dariko commented:
@Akasurde I repeat: I think I have addressed the problems you mentioned, have i missed something?

gesellix commented:
@chouseknecht @bcoca maybe @Akasurde doesn't read our comments, maybe you can help to make this PR pass?

LuisGerardoAngel commented:
@Akasurde Is this up for the next release?

Akasurde commented:
@LuisGerardoAngel Sorry I missed comments under the pile of notifications.

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/python
+
+# This file is part of Ansible
```
Replace line 3 to 16 with 

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Copyright: (c) 2017, Ansible Project
```

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'metadata_version': '1.0'}
```
s/1.0/1.1/

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+    Manage docker stacks using the 'docker stack' command
+    on the target node
+    (see examples)
+version_added: "2.4"
```
s/2.4/2.5/

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+        required: false
+        default: ""
+        description:
+        -   String containing the yaml definition of the stack
```
Fullstop at end of sentence.

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+                docker_stack_spec_diff=str(before_after_differences))
+
+    else:
+        rc, out, err = module.run_command(["docker",
```
Please use 

```
docker_bin = module.get_bin_path('docker', required=True)
rc, out, err = module.run_command([docker_bin, ...
```

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+                             err=err)
+
+        module.exit_json(changed=True, msg=out, err=err)
+
```
```if __name__ == "__main__":```

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+
+
+def docker_stack_deploy(module, stack_name, compose_file):
+    command = ["docker", "stack", "deploy"]
```
```docker_bin = module.get_bin_path('docker', required=True)```

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+
+
+def docker_service_inspect(module, service_name):
+    rc, out, err = module.run_command(["docker",
```
```docker_bin = module.get_bin_path('docker', required=True)```

Akasurde commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,224 @@
+
+
+def docker_stack_services(module, stack_name):
+    rc, out, err = module.run_command(["docker",
```
```docker_bin = module.get_bin_path('docker', required=True)```

Akasurde reviewed (changes requested):
Sorry missed due to lots of notifications.

dariko committed:
update header

dariko committed:
refresh metadata versions

dariko committed:
end sentences in fullstop

dariko committed:
use module.get_bin_path to find docker binary path

dariko committed:
execute main only if run directly

dariko commented:
I need some help with the errors returned by CI:
- I dot't know what this means:
  `````
  == Missing from __future__ import (absolute_import, division, print_function) ==
  ./lib/ansible/modules/cloud/docker/docker_stack.py
  `````
- There are failures related to the unavailability of the jsondiff module:
  `````
  failure: The test `ansible-test sanity --test import --python 2.6` failed with the following error: 
  lib/ansible/modules/cloud/docker/docker_stack.py:95:0: ImportError: No module named jsondiff
  `````
  but the `jsondiff` requirement is explicit in the module metadata. Should I add it somewhere else?
- Lastly, I suppose the error:
  `````
  lib/ansible/modules/cloud/docker/docker_stack.py:0:0: E321 Exception attempting to import module for argument_spec introspection
  `````
  is a consequence of the missing jsondiff module.

thanks

mattclay commented:
@dariko Take a look at the help doc links. They're the `[?]` in the message posted by ansibot.

More specifically:

- For an example of a `__future__` import for a module, take a look at: https://github.com/ansible/ansible/blob/db61f8f9673bb03a34ff55e3d8d31b7225c6f47d/lib/ansible/modules/system/ping.py#L8
- For an example of handling import errors in modules, take a look at: https://github.com/ansible/ansible/blob/ba22514d479777967da292d12ac542c8c727af2e/lib/ansible/modules/source_control/gitlab_group.py#L104-L108 https://github.com/ansible/ansible/blob/ba22514d479777967da292d12ac542c8c727af2e/lib/ansible/modules/source_control/gitlab_group.py#L166-L167

ushuz committed:
Add with_registry_auth option

https://github.com/dariko/ansible-module-docker_stack/pull/2

ushuz committed:
Add resolve_image option

ushuz committed:
Fix ansible boilerplate

ushuz committed:
Conform to ansible python imports guideline

https://github.com/ansible/ansible/blob/4956bfd4c4b377210426dc9d83ca1e1515f6417f/CODING_GUIDELINES.md#python-imports

ushuz commented:
@dariko I've addressed some complaints of @ansibot in https://github.com/dariko/ansible/pull/2

phoenix741 commented:
Is it possible to add the option to activate the parameters : --with-registry-auth that i use to authenticate on a private registry ?

Thanks

mattclay commented:
@dariko Can you include integration tests for this new module?

kireevco commented:
Is this a direct replacement of `docker_swarm_mode` module?

dariko commented:
@kireevco I can't find a module named `docker_swarm_mode`.
There is a `docker_swarm` module in this PR: https://github.com/ansible/ansible/pull/38280 which seems about managing a swarm clusters and member nodes, not stacks running on the cluster.

kireevco commented:
@dariko sorry, i meant your `docker_swarm_service` PR https://github.com/ansible/ansible/pull/19229

dariko commented:
@kireevco The `docker_swarm_service` module manages a  single service while this one manages a stack (a collection of services).
I seem to remember there were some limitations on stack managed services compared to standalone services, but they might have been overcome.

kireevco commented:
Does this module currently support (tested and is encouraged to use for) deployment of stacks to docker swarm mode?

Thanks.

kireevco commented:
@dariko so bottom line, those are intended to be kept side by side? https://github.com/ansible/ansible/pull/24588 and https://github.com/ansible/ansible/pull/19229 ?

dariko commented:
@kireevco there is no relation between the modules proposed in #24588 and #19229. They do not require each other.

MadDaelim commented:
What with multiple compose files for docker stack deploy? Feature was added in docker 18.03.0 https://github.com/docker/cli/pull/569

dariko commented:
@MadDaelim thanks, I missed that.

I think to support this feature the module interface should change: I could replace the `compose_{file,yaml}` parameters with a new one (named... `compose`? `compose_defs`?) which would be a list whose elements could be strings (treated as the old `compose_file`) or dictionaries (mapping to the old `compose_yaml`).  
ex:
`````
-   name: deploy stack1 from my_generic_compose.yml + customization
    docker_stack:
        state: present
        name: stack1
        compose:
        -   my_generic_compose.yml
        -   version: '3'
            services:
                myservice:
                    # change image
                    image: alpine:latest
        #[...] other list elements
`````
Anyone see any problem with this approach?

dariko committed:
correct documentation for CI

dariko committed:
Replace `compose_yaml`, `compose_file` parameters with `compose`.

Since version 18.03.0 docker allows for a stack to be defined using
multiple compose files.

This commit changes the module interface to allow this new feature to
be used, by replacind the `compose_{file,yaml}` parameters with a new
`compose` one which is expected to contain a list of:
- compose file paths (on the target system)
- compose yaml definitions

dariko commented:
@mattclay I'm gonna read the docs about integration testing

dariko committed:
correct idempotency for state==absent

dariko committed:
add absent_retries, absent_retries_interval parameters

MadDaelim commented:
@dariko looks good for me

dariko committed:
.

dariko committed:
output docker_stack_spec_diff as json

dariko committed:
update documentation, cleanup pep8 problems

dariko committed:
verify compose parameter is a valid list

dariko committed:
basic integration test

dariko committed:
correct indentation

dariko committed:
install jsondiff and pyyaml

dariko committed:
Skip integration tests on fedora<26

On Fedora 24 the last version of docker ce available is too old to
initialize a stack using the module docker_stack.
Running this task:
`````
- name: Create a Swarm cluster
  docker_swarm:
    state: present
    advertise_addr: "{{ansible_default_ipv4.address}}"
`````
fails with
`````
TASK [docker_stack : Create a Swarm cluster]
***********************************
task path:
/root/ansible/test/integration/targets/docker_stack/tasks/test_stack.yml:8
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo
/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651 `" && echo
ansible-tmp-1535925972.57-142477246776651="` echo
/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651 `" ) &&
sleep 0'
Using module file
/root/ansible/lib/ansible/modules/cloud/docker/docker_swarm.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-103nuR5B6/tmpIEXODD TO
/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x
/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/
/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py
&& sleep 0'
<127.0.0.1> EXEC /bin/sh -c
'/tmp/ansible-test-coverage-MYVXzL/coverage/injector.py
/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py
&& sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r
/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/ >
/dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File
"/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py",
line 113, in <module>
    _ansiballz_main()
  File
"/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py",
line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File
"/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py",
line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_docker_swarm_payload_24vojG/__main__.py", line 525,
in <module>
  File "/tmp/ansible_docker_swarm_payload_24vojG/__main__.py", line 520,
in main
  File "/tmp/ansible_docker_swarm_payload_24vojG/__main__.py", line 306,
in __call__
  File "/tmp/ansible_docker_swarm_payload_24vojG/__main__.py", line 329,
in init_swarm
  File "/tmp/ansible_docker_swarm_payload_24vojG/__main__.py", line 389,
in __update_swarm
  File "/tmp/ansible_docker_swarm_payload_24vojG/__main__.py", line 357,
in __update_spec
KeyError: 'KeepOldSnapshots'

fatal: [testhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File
\"/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py\",
line 113, in <module>\n    _ansiballz_main()\n  File
\"/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py\",
line 105, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path,
ANSIBALLZ_PARAMS)\n  File
\"/root/.ansible/tmp/ansible-tmp-1535925972.57-142477246776651/AnsiballZ_docker_swarm.py\",
line 48, in invoke_module\n    imp.load_module('__main__', mod, module,
MOD_DESC)\n  File
\"/tmp/ansible_docker_swarm_payload_24vojG/__main__.py\", line 525, in
<module>\n  File
\"/tmp/ansible_docker_swarm_payload_24vojG/__main__.py\", line 520, in
main\n  File \"/tmp/ansible_docker_swarm_payload_24vojG/__main__.py\",
line 306, in __call__\n  File
\"/tmp/ansible_docker_swarm_payload_24vojG/__main__.py\", line 329, in
init_swarm\n  File
\"/tmp/ansible_docker_swarm_payload_24vojG/__main__.py\", line 389, in
__update_swarm\n  File
\"/tmp/ansible_docker_swarm_payload_24vojG/__main__.py\", line 357, in
__update_spec\nKeyError: 'KeepOldSnapshots'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
`````

This commit allows the tests to be on Fedora only if version 26 or
later, whose docker-ce repositories contain more recent versions.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,304 @@
+        required: false
+        default: false
+        description:
+        -   If true will add the `--prune` option to the `docker stack deploy` command.
```
You cannot use backticks for formatting here. Use `C(--prune)` and `C(docker stack deploy)` instead.

felixfontein replied:
(There are some more places where you use backticks; I won't mark them all.)

dariko replied:
Thanks, I'm changing them

felixfontein commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,304 @@
+    if state == 'present':
+        if not compose:
+            module.fail_json(msg=("compose parameter must be a list "
+                                  "containing at least an element"))
```
maybe 'an' -> 'one'?

dariko replied:
That was the idea... I'm changing it

felixfontein commented on lib/ansible/modules/cloud/docker/docker_stack.py:
```diff
@@ -0,0 +1,304 @@
+            after_stack_services = docker_stack_inspect(module, name)
+
+        finally:
+            for temp_file in temp_files:
```
`finally` blocks are not executed if `module.fail_json` is used (`fail_json()` calls `sys.exit()`)!

To make sure that a file is deleted, use `module.add_cleanup_file(filename)` after creating it.

dariko replied:
Ok, I removed the whole try..finally: using `module.add_cleanup_file` it is now unnecessary

felixfontein commented on test/integration/targets/docker_stack/tasks/test_stack.yml:
```diff
@@ -0,0 +1,99 @@
+- name: assert state=absent idempotency
+  assert:
+    that:
+    - not output is changed
```
why not `output is not changed`?

dariko replied:
This test is verifying the module does **not** report the task as changed when run with `state==absent` and a name referring to a stack which does not exist (in this case, it should have been just deleted).

felixfontein replied:
`not A is B` is equivalent to `A is not B`. I find the second form easier to read, as it is more like natural language.

dariko replied:
Sorry, I misunderstood your previous comment.  
I changed the test as you suggested.

felixfontein replied:
No problem :) Thanks!

felixfontein reviewed (commented):
I've never used `docker stack`, so I can't comment on the main part of the module, but I found some little things (and one not so little) while looking over it :)

dariko committed:
correct syntax and formatting

dariko committed:
Use AnsibleModule.add_cleanup_file to ensure temporary files are removed

dariko committed:
clarify assertion tests + typo

felixfontein reviewed (approved):
shipit

gundalow commented:
Merged into `devel` for release in Ansible 2.8.
Thank you all for the reviews

If you search for `docker` on https://ansible.sivel.net/pr/byfile.html there are a few other PRs open. Reviews on those would be very welcome.

If you'd like to become maintainers of Docker (so you are notified of issues & PR) and have voting rights, we'd welcome you to raise a PR to add yourself to https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml

Any questions feel free to ping me on `#ansible-devel`

mattclay commented:
The tests in this PR are conflicting with the tests for docker_swarm:

https://app.shippable.com/github/ansible/ansible/runs/84464/50/tests

I'm looking into the issue.

mattclay commented:
Fix for the issue: https://github.com/ansible/ansible/pull/45807

mattclay commented:
The "fix" has been merged. It's really a work-around at this point which will hopefully resolve the issues with tests failing.

I ended up adding a task to the end of the docker_stack test to leave the swarm. This should prevent future docker tests from having issues. However, doing this exposed another problem: it wasn't possible to leave the swarm if the stack had been updated. I don't know why this happens, so I ended up disabling that part of the test so the rest of the test could still execute.

Hopefully someone has time to investigate further and fix the underlying issue.
